app/src/org/commcare/android/nfc/NdefRecordUtil.java

```diff
@@ -0,0 +1,133 @@
+        if (Arrays.equals(record.getType(), NdefRecord.RTD_TEXT)) {
+            if (Arrays.asList(userSpecifiedTypes).contains(("text"))) {
+                return new Pair<>(readValueFromTextRecord(record), true);
+            } else {
+                return new Pair<>(READ_ERROR_MISMATCH_KEY, false);
+            }
+        } else {
+            return new Pair<>(READ_ERROR_UNSUPPORTED_KEY, false);
+        }
+    }
+
+    private static Pair<String,Boolean> handleExternalTypeRecord(NdefRecord record,
+                                                                 String[] userSpecifiedTypes,
+                                                                 String domainForExternalTypes) {
+        String typeAsString = new String(record.getType(), UTF8_CHARSET);
+        if (recordTypeIsExpected(typeAsString, userSpecifiedTypes, domainForExternalTypes)) {
+            return new Pair<>(new String(record.getPayload(), UTF8_CHARSET), true);
+        } else {
+            return new Pair<>(READ_ERROR_MISMATCH_KEY, false);
+        }
+    }
+
+    private static boolean recordTypeIsExpected(String recordTypeAsString,
+                                                String[] userSpecifiedTypes,
+                                                String domainForExternalTypes) {
+        for (String type : userSpecifiedTypes) {
+            type = domainForExternalTypes + ":" + type;
+            if (type.equals(recordTypeAsString)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static String readValueFromTextRecord(NdefRecord textTypeRecord) {
+        byte[] fullPayload = textTypeRecord.getPayload();
+        // The payload includes a prefix denoting the language, so we need to parse that off
+        int langBytesLength = fullPayload[0]; // status byte
+        int lengthOfPrefix = langBytesLength + 1; // add 1 for the status byte itself
+        byte[] payloadWithoutLang = Arrays.copyOfRange(fullPayload, lengthOfPrefix, fullPayload.length);
+        return new String(payloadWithoutLang, UTF8_CHARSET);
+    }
+
+    protected static NdefRecord createNdefRecord(String userSpecifiedType,
+                                                 String userSpecifiedDomain,
+                                                 String payloadToWrite) {
+        if (isCommCareSupportedWellKnownType(userSpecifiedType)) {
+            return createWellKnownTypeRecord(userSpecifiedType, payloadToWrite);
+        } else {
+            return createExternalRecord(userSpecifiedType, userSpecifiedDomain, payloadToWrite);
+        }
+    }
+
+    protected static boolean isCommCareSupportedWellKnownType(String type) {
+        // For now, the only "well known type" we're supporting is NdefRecord.RTD_TEXT, which
+        // users should encode in their configuration by specifying type "text"
+        return "text".equals(type);
```

Comment:
We've discussed how WTO might come back and ask for some encrypted/encoded format, but all of this `type` checking has me thinking YAGNI.

Reply:
Are you referring to not supporting the custom record types at all? If so, I would really rather not do that. I think there is a high probability that 1 of the next few projects that want to use this will require something like the custom types (and the work is already done to include it in a way that I feel does not add an unreasonable complexity burden). I also think there is a non-trivial chance that 2 months from now, World Vision will decide there is some reason they actually do need this.